src/sender.rs

```diff
@@ -1,85 +1,499 @@
 use crate::*;
-use alloc::sync::Arc;
-use core::{future::Future, task::Poll};
-use futures_micro::poll_fn;
+#[cfg(feature="async")]
+use core::{future::Future, pin::Pin, task::{Context, Poll}};
+use core::ops::Deref;
+use core::sync::atomic::Ordering::Relaxed;
 
-/// The sending half of a oneshot channel.
+#[derive(Debug,Eq,PartialEq)]
+/// The reason we couldn't send a message. And our value back.
+pub enum SendError<T> {
+    Closed(T),
+    Existing(T),
+}
```

Comment:
imo I would split this into a `struct` + `enum` as such:
```rust
#[derive(Clone,Copy,Debug,Eq,PartialEq)]
/// The reason why we couldn't send a message.
pub enum SendErrorKind {
    Closed,
    Existing,
}
#[derive(Debug,Eg,PartialEq)]
pub struct SendError<T> {
    pub kind: SendErrorKind,
    pub item: T,
}
```

Reply:
It feels a bit... anti-minimalistic?

Reply:
albeit being a bit anti-minimalistic, I think it would keep most of the convenience, all while reducing the complexity in the trait implementations, and maybe even making some trait implementations (e.g. `Deref`) unnecessary. I think that abstraction would also represent better what that type wants to convey, because the `item` is completely independent of the "reason"/`kind`.